packages/components/src/confirm-dialog/README.md

```diff
@@ -0,0 +1,94 @@
+  * It will be automatically open (displayed) upon mounting;
+  * It will automatically close itself when the `x`, `Cancel`, `Confirm` or overlay is clicked;
+  * `onCancel` is not mandatory but can be passed. Even if passed, the dialog will still be able to close itself.
+
+Activating this mode is as simple as omitting the `isOpen` prop. The only mandatory prop, in this case, is the `onConfirm` callback.
+
+```jsx
+import {
+  __experimentalConfirmDialog as ConfirmDialog
+} from '@wordpress/components';
+
+function Example() {
+  return (
+    <ConfirmDialog onConfirm={ () => console.debug(' Confirmed! ') }>
+  )
+}
+```
+
+### Controlled mode
+
+Let the parent component control when the dialog is open/closed. It's activated when a boolean value is passed to `isOpen`:
+  * It will not be automatically closed. You need to let it know when to open/close by updating the value of the `isOpen` prop;
+  * Both `onConfirm` and the `onCancel` callbacks are mandatory props in this mode;
+  * You'll want to update the state that controls `isOpen` by updating it from the `onCancel` and `onConfirm` callbacks.
+
+
+```jsx
+import {
+  __experimentalConfirmDialog as ConfirmDialog
+} from '@wordpress/components';
+
+function Example() {
+  const [ isOpen, setIsOpen ] = useState( true );
+
+  const handleConfirm = () => {
+    console.debug( 'Confirmed!' );
+    setIsOpen( false );
+  }
+
+  const handleCancel = () => {
+    console.debug( 'Cancelled!' );
+    setIsOpen( false );
+  }
+
+  return (
+    <ConfirmDialog isOpen={ isOpen } onConfirm={ handleConfirm } onCancel={ handleCancel }>
+  )
+}
+```
+
+### Unsupported: Multiple instances
+
+Multiple `ConfirmDialog's is an edge case that's currently not officially supported by this component. At the moment, new instances will end up closing the last instance due to the way the `Modal` is implemented (more specifically, because of the use of the `onFocusOutside` to detect a blur in order to close it). This will end up automatically cancelling the previous instance, which might or might not be what you want.
+
+Suppose you need to handle multiple confirmations without discarding other instances. In that case, it might be better to have a singleton wrapper component that provides a context-based API to trigger the dialog, keeping track of multiple instances. Here's an [example](https://github.com/WordPress/gutenberg/pull/34153#issuecomment-908342367).
+
+### Props
+
+### `isOpen`: `boolean`
+
+Defines if the dialog is open (displayed) or closed (not rendered/displayed). It also implicitly toggles the controlled mode if set or the uncontrolled mode if it's not set.
+
+### `onConfirm`: `( event: SyntheticEvent ) => void`
+
+- Required: Yes
+
+The callback that's called when the user confirms. A confirmation can happen when the `OK` button is clicked or when `Enter` is pressed.
+
+### `onCancel`: `(event: SyntheticEvent ) => void`
+
+- Required: Yes if `isOpen` is set, No if `isOpen` is not set
```

Comment:
> - Required: Yes if `isOpen` is set, No if `isOpen` is not set

Is there any way we can improve this situation, so that we don't have conditionally required props? Maybe the best solution is to make it always required?

Or to expose two different components, one controlled and one uncontrolled?

Reply:
We could also try experimenting with TypeScript types, something like

```ts
type typeA = {
  isOpen?: boolean;
  onCancel?: (event: SyntheticEvent ) => void;
};

// `onCancel` is required if `isOpen` is set
type typeB = {
  isOpen: boolean;
  onCancel: (event: SyntheticEvent ) => void;
};

type union = typeA | typeB;
```

Not sure.

Reply:
Yeah, this is the part I want the most feedback on. I came up with this API as I developed the component to look and behave like the native `confirm` component (and be easy and straightforward to use), while still allowing it to be used as a controlled component. I think it mainly depends on how we plan on consuming this component:

* If we think it's fine to just use it uncontrolled inline in the components that need it, then yes, we should keep this or break into two components as you said;
* if we'll instead add more infrastructure around it (like what Diego suggested [here](https://github.com/WordPress/gutenberg/pull/34153#issuecomment-908342367)) maybe it doesn't make sense to keep the uncontrolled aspect.